### ISSUE TITLE
Class cast exception when data binding to a custom collection class

### ISSUE THREAD

**Issue by lexspoon** [FasterXML/jackson-databind#1133]
I have a data model that I could data-bind automatically using Jackson 2.2.0, but that has stopped working as of 2.6.0 and 2.6.4. Sorry if this is a duplicate report, but I thought you could at least use it as a test case even if so.

The custom collection class has a definition like this:
```
public final class Seq<T> extends AbstractList<T> {
	@JsonCreator
	public static <T> Seq<T> of(Iterable<T> elems) { ... }
	...
}
```

Here's a typical use of this class in another data model:
```
public class BuildGraph {
	@JsonCreator
	public BuildGraph(@JsonProperty("environ") Seq<String> environ,
			@JsonProperty("localenv") Seq<String> localenv,
			@JsonProperty("rules") Seq<BuildRule> rules) { ... }
	...
```

With Jackson 2.6.0 or 2.6.4, trying to deserialize into this model gives me the following class cast exception:
```
java.util.LinkedHashMap cannot be cast to com.semmle.bcache.BuildRule
```

I've shrunk my code base to a small skeleton of code that still reproduces the error. I don't see a way to attach things to a GitHub issue ticket, so I pushed it to a new repository:

https://github.com/lexspoon/bug-jackson-data-bind-seq

To run the test, just run `./run.sh`. You can edit the file to change which version of Jackson gets used.

I believe this should work, or at least not yield a class cast exception. However, I am also interested if there is anyway to annotate the Seq class so that this will work. My current workaround is to accept java.util.List in constructor parameters and then manually convert it to a Seq in the constructor, but that's not nearly as nice as if Jackson can data bind Seq automatically.

**Comment by cowtowncoder** [comment]
Quick note on this one: since type resolution has completely rewritten between 2.6 and 2.7, it could be that 2.7[.1] works better than 2.6.

If not, I hope to have a look at this when I get through other open issues (I have bit of backlog); thank you for providing the reproduction, that should help a lot in figuring the issue.

**Comment by lexspoon** [comment]
Good news--the problem is fixed as of 2.7.0.

I'll leave this open in case you want to add the data model as a test case. It uses a custom Java collection subclass, so it should be a good exercise of the generics logic in jackson-databind.

**Comment by cowtowncoder** [comment]
@lexspoon Thanks, good idea.
